pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+	///   need to resolve the delta between both swap order amounts and update
+	///   the states accordingly.
+	#[allow(clippy::type_complexity)]
+	fn handle_concurrent_swap_orders(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		swap_order_id: T::TokenSwapOrderId,
+		reason: TokenSwapReason,
+	) -> Result<
+		(
+			Option<Swap<T::Balance, T::CurrencyId>>,
+			Option<InvestState<T::Balance, T::CurrencyId>>,
+			Option<RedeemState<T::Balance, T::CurrencyId>>,
+		),
+		DispatchError,
+	> {
+		let last_reason = ForeignInvestmentInfo::<T>::get(swap_order_id)
+			.ok_or(Error::<T>::InvestmentInfoNotFound)?
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		// Exit early if both reasons match, i.e. we would not override any opposite
+		// swap order
+		if last_reason == reason {
+			return Ok((None, None, None));
+		}
+
+		// Read states from storage and determine amounts
+		let invest_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+		let redeem_state = RedemptionState::<T>::get(who, investment_id).unwrap_or_default();
+		let invest_swap_amount = invest_state
+			.get_active_swap()
+			.map(|s| s.amount)
+			.unwrap_or_default();
+		let redeem_swap_amount = redeem_state
+			.get_active_swap()
+			.map(|s| s.amount)
+			.unwrap_or_default();
+		let resolved_amount = invest_swap_amount.min(redeem_swap_amount);
+		// safe because max >= min, equals zero if both amounts equal
+		let swap_amount_opposite_direction =
+			invest_swap_amount.max(redeem_swap_amount) - resolved_amount;
+
+		// Determine new invest state
+		let new_invest_state = match invest_state {
+			// As redeem swap can only be into foreign currency, we need to delta on the opposite
+			// swap directions
+			InvestState::ActiveSwapIntoPoolCurrency { swap } => {
+				if invest_swap_amount > redeem_swap_amount {
+					Some(
+						InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+							swap: Swap {
+								amount: swap_amount_opposite_direction,
+								..swap
+							},
+							invest_amount: resolved_amount,
+						},
+					)
+				} else {
+					Some(InvestState::InvestmentOngoing {
+						invest_amount: resolved_amount,
+					})
+				}
+			}
+			// Same as above except for the base investment amount which is incremented
+			InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+				swap: invest_swap,
+				invest_amount,
+			} => {
+				if invest_swap_amount > redeem_swap_amount {
+					Some(
+						InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+							swap: Swap {
+								amount: swap_amount_opposite_direction,
+								..invest_swap
+							},
+							invest_amount: invest_amount.ensure_add(resolved_amount)?,
+						},
+					)
+				} else {
+					Some(InvestState::InvestmentOngoing {
+						invest_amount: invest_amount.ensure_add(resolved_amount)?,
+					})
+				}
+			}
+			// We must not alter the invest state if there is no active pool currency swap
+			_ => None,
```

Comment:
I think this is an issue, if we do not want to alter the invest state, we need to return here `invest_state`. If we return `None`, then storage will be overwritten with `None`, removing the state.

Reply:
I am really sorry about the complexity and bad readability of `handle_concurrent_swap_orders` function in combination with the state apply functions.

Returning `None` here is correct in the current design. The whole idea of `handle_concurrent_swap_orders` is that if it returns `Some(state)` as part of the return-tuple `(Option<Swap>, Option<InvestState>, Option<RedeemState>)`, this reflects that this particular state has changed as a result of executing `handle_concurrent_swap_orders`. So returning `None` here implies no state change occured.

See `handle_swap_order` where the potential state updates are executed: https://github.com/centrifuge/centrifuge-chain/blob/f226e64f343035352ef084978772b0ff7e6ce122/pallets/foreign-investments/src/impls/mod.rs#L711-L759

Reply:
No worries about anything!

Just the code you've shared, in line `730`, the `current_invest_state` is updated in case it's `None`, or I'm not seeing it well.

Reply:
I think I've pointed you to the wrong `None`, I think is here:

```rust
		// Exit early if both reasons match, i.e. we would not override any opposite
		// swap order
		if last_reason == reason {
			return Ok((None, None, None));
		}
```

So, if the reason is the same (i.e. I increase a previous new investment), the result is `None`, and a possible existent `ActiveSwapIntoPoolCurrency` state is removed instead of being incremented with the new amount.

Reply:
Again, apologies for the complex interactions. If both swap reasons are the same, the `apply_invest_state_transition` or respectively `apply_redeem_state_transition` already updated the `InvestmentState`/`RedemptionState` entry. 

In your exemplary case this would be handled here:
https://github.com/centrifuge/centrifuge-chain/blob/f226e64f343035352ef084978772b0ff7e6ce122/pallets/foreign-investments/src/impls/mod.rs#L371-L376

Reply:
Again, no worries! You're working on a very complex problem!

My question came from [this test](https://github.com/centrifuge/centrifuge-chain/pull/1509/files#diff-501ed038c8b63b0bf86c800dcd90838a270db273656645b5cb4def9d533199eeR139) (last `assert_eq!` I obtained `None`). But after rebasing your changes where you do not use `Option` for `InvestState` the issue I had with this seems that no longer applies.

Let me re-check how it would work now, but I think now it's solved.

Reply:
Did you re-check the test?

Reply:
Yes (sorry I didn't update)! The test now contains the correct `InvestState` instead of `None`, 👍🏻